lib/tools/system-calls.js

```diff
@@ -75,4 +76,112 @@ systemCallMethods.getBinaryFromPath = async function(binaryName) {
   }
 };
 
+systemCallMethods.getConnectedDevices = async function () {
+  log.debug("Getting connected devices...");
+  try {
+    let {stdout} = await exec(this.adb.path, ['devices']);
+    if (stdout.toLowerCase().indexOf("error") !== -1) {
+      log.error(stdout);
+      throw new Error(stdout);
+    } else {
+      let devices = [];
+      for (let line of stdout.split("\n")) {
+        if (line.trim() !== "" &&
+            line.indexOf("List of devices") === -1 &&
+            line.indexOf("* daemon") === -1 &&
+            line.indexOf("offline") === -1) {
+          let lineInfo = line.split("\t");
+          // state is either "device" or "offline", afaict
+          devices.push({udid: lineInfo[0], state: lineInfo[1]});
+        }
+      }
+      log.debug(`${devices.length} device(s) connected`);
+      return devices;
+    }
+  } catch (e) {
+    throw new Error(`Error while getting connected devices. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.getDevicesWithRetry = async function (timeoutMs = 20000) {
+  let start = Date.now();
+  log.debug("Trying to find a connected android device");
+  let getDevices = async () => {
+    if ((Date.now() - start) > timeoutMs) {
+      throw new Error("Could not find a connected Android device.");
+    }
+    try {
+      let devices = await this.getConnectedDevices();
+      if (devices.length < 1) {
+        log.debug("Could not find devices, restarting adb server...");
+        await this.restartAdb();
+        // cool down
+        await sleep(200);
+        await getDevices();
+      }
+      return devices;
```

Comment:
but just calling `await getDevices()` recursively doesn't get that value back to the original calling function ever, right? this probably needs some heavy unit tests since there's so much actual logic. on first pass I think that would show up a bug. I think this block should probably look like:

```js
      let devices = await this.getConnectedDevices();
      if (devices.length < 1) {
        log.debug("Could not find devices, restarting adb server...");
        await this.restartAdb();
        // cool down
        await sleep(200);
        devices = await getDevices();
      }
      return devices;
```

but i could be wrong